src/theory/tools/wp/bir_htScript.sml

```diff
@@ -0,0 +1,614 @@
+        FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+        FULL_SIMP_TAC (std_ss++holBACore_ss) []
+      ]
+    ]
+  ],
+
+  (* CJmp *)
+  FULL_SIMP_TAC std_ss [bir_exec_stmtE_def,
+                        bir_exec_stmt_cjmp_def] >>
+  Cases_on `bir_dest_bool_val (bir_eval_exp b st.bst_environ)` >- (
+    FULL_SIMP_TAC (std_ss++holBACore_ss) [bir_state_set_failed_def]
+  ) >>
+  Cases_on `x` >> (
+    FULL_SIMP_TAC std_ss [bir_exec_stmt_jmp_def]
+  ) >>> (TACS_TO_LT [
+    Cases_on `bir_eval_label_exp b0 st.bst_environ` >- (
+      FULL_SIMP_TAC (std_ss++holBACore_ss)
+                    [bir_state_set_failed_def]
+    ),
+
+    Cases_on `bir_eval_label_exp b1 st.bst_environ` >- (
+      FULL_SIMP_TAC (std_ss++holBACore_ss)
+                    [bir_state_set_failed_def]
+    )
+  ]) >> (
+    Cases_on `x` >| [
+      FULL_SIMP_TAC (std_ss++holBACore_ss)
+                    [bir_state_set_failed_def,
+                     bir_exec_stmt_jmp_to_label_def] >>
+      Cases_on `MEM (BL_Label s) (bir_labels_of_program prog)` >| [
+        FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+        FULL_SIMP_TAC (std_ss++holBACore_ss) []
+      ],
+
+      FULL_SIMP_TAC (std_ss++holBACore_ss)
+                    [bir_exec_stmt_jmp_to_label_def] >>
+      Cases_on `MEM (BL_Address b')
+                    (bir_labels_of_program prog)` >| [
+        FULL_SIMP_TAC (std_ss++holBACore_ss) [],
+
+        FULL_SIMP_TAC (std_ss++holBACore_ss) []
+      ]
+    ]
+  ),
+
+  (* Halt *)
+  FULL_SIMP_TAC (std_ss++holBACore_ss) [bir_exec_stmtE_def,
+                                        bir_exec_stmt_halt_def]
+]
+);
+
+(* TODO: Move this elsewhere... *)
+val bir_exec_stmtsB_exists =
+  store_thm("bir_exec_stmtsB_exists",
+  ``!stmtsB l c st.
+      ?l' c' st'.
+        bir_exec_stmtsB stmtsB (l,c,st) = (l', c', st')``,
+
+Induct_on `stmtsB` >- (
+  FULL_SIMP_TAC std_ss [bir_exec_stmtsB_def]
+) >>
+REPEAT STRIP_TAC >>
+FULL_SIMP_TAC std_ss [bir_exec_stmtsB_def] >>
+Cases_on `bir_state_is_terminated st` >- (
+  FULL_SIMP_TAC std_ss []
+) >>
+ASSUME_TAC (SPECL [``h:'a bir_stmt_basic_t``, ``st:bir_state_t``]
+                  bir_exec_stmtB_exists) >>
+FULL_SIMP_TAC std_ss [LET_DEF]
+);
+
+val bir_block_not_assume_never_assviol =
+  store_thm("bir_block_not_assume_never_assviol",
+  ``!prog bl st l' c' st'.
+    bir_block_has_no_assumes bl ==>
+    (st.bst_status <> BST_AssumptionViolated) ==>
+    (bir_exec_block prog bl st = (l', c', st')) ==>
+    (st'.bst_status <> BST_AssumptionViolated)``,
+
+FULL_SIMP_TAC std_ss [bir_block_has_no_assumes_def,
+                      bir_exec_block_def] >>
+REPEAT STRIP_TAC >>
+IMP_RES_TAC bir_stmtsB_not_assume_never_assviol >>
+subgoal `?l_test c_test st'_test.
+	   bir_exec_stmtsB bl.bb_statements ([],0,st) =
+             (l_test, c_test, st'_test)` >- (
+  FULL_SIMP_TAC std_ss [bir_exec_stmtsB_exists]
+) >>
+FULL_SIMP_TAC std_ss [LET_DEF] >>
+PAT_X_ASSUM ``!st' l' l c' c. blah``
```

Comment:
IMO you should use `_` instead of `blah`. That's what I infer from https://hol-theorem-prover.org/kananaskis-12.release.html#incompatibilities , and I find it a bit cleaner.

Reply:
yes, sound good

Reply:
Yep, this is cleaner for sure, and it agrees with the wildcard operator in SML. I will change this.

Can you quote the text which suggests usage of _ over other dummy strings in the link?

Reply:
> Can you quote the text which suggests usage of _ over other dummy strings in the link?

Sure! I wanted to link the following, even if it kinda doesn't apply here (emphasis mine):

> **The behaviour of the quoting versions of the tactics is also slightly different**: **they will always respect names** that occur both in the pattern and in the goal. Again, this is for consistency with similar functions such as `qspec_then`. This means, for example, that `qpat_assum 'a < b'` **will fail** if the actual theorem being matched is something `c < f a`. (This is because the pattern and the goal share the name `a`, so that the pattern is implicitly requiring the first argument to `<` to be exactly `a`, which is not the case.) This example would have previously worked if there was exactly one assumption with `<`. **The fix in cases like this is to use more underscores in one’s patterns**.